### PR TITLE
WIP: Bugfix: fix issue 145 compile on VS 2022

### PR DESCRIPTION
**Pull request by burnpanck**
This PR provides a workaround around a compiler bug, which seems to have appeared in VS 2022. It appears that the bug is only triggered with `/permissive-`, which is enabled by default when using `/std:c++20`. As I was unable to make CMAKE set the former, I simply set the latter in the accompanying CI tests.

In this version of the PR, the workaround is hidden behind a `FOONATHAN_SFINAE_WORKAROUND` guard, which for now is implemented unconditionally, as I haven't fully determined what versions and compiler flag combinations really trigger the issue. Since you have used `std::declval` in other places within `FOONATHAN_SFINAE`, maybe the workaround could actually be used unconditionally everywhere, and the guard removed? Let me know.

### CODE REVIEW

File: .github/workflows/main_ci.yml [hunk 1]
```diff
@@ -2,7 +2,7 @@ name: Main CI
 
 on:
   push:
-    branches: [main]
+    branches: [main, bugfix/fix-issue-145-compile-on-msvc-2022]
```
Comment (foonathan) [inline .github/workflows/main_ci.yml:5]: I'm assuming you just intend this to be temporary for testing?

File: .github/workflows/main_ci.yml [hunk 2]
```diff
@@ -81,8 +81,9 @@ jobs:
       matrix:
         build_type: [Debug, Release]
         sharedlibs: [OFF, ON]
+        std: [14, 17, 20]
 
-    runs-on: windows-2019
+    runs-on: windows-latest
```
Comment (foonathan) [inline .github/workflows/main_ci.yml:86]: Please hardcode the specific windows version you want; GitHub updates it occasionally which changes the generator.

File: include/foonathan/memory/allocator_storage.hpp [hunk 3]
```diff
@@ -126,7 +126,11 @@ namespace foonathan
                 FOONATHAN_REQUIRES(
                     (!std::is_base_of<allocator_storage, typename std::decay<Alloc>::type>::value))>
             allocator_storage(Alloc&& alloc,
+#if FOONATHAN_SFINAE_WORKAROUND
```
Comment (foonathan) [inline include/foonathan/memory/allocator_storage.hpp:129]: Conditional compilation is not necessary; you can use declval unconditionally.

File: include/foonathan/memory/config.hpp [hunk 4]
```diff
@@ -43,6 +43,11 @@
 #endif
 #endif
 
+// SFINAE workaround for MSVC 19.33
+#ifndef FOONATHAN_SFINAE_WORKAROUND
```
Comment (foonathan) [inline include/foonathan/memory/config.hpp:47]: As above; this is unnecessary.

**Review by foonathan — CHANGES_REQUESTED** [review]
Thank you; see comments.

**Comment by foonathan** [comment]
Ignore the MacOS failures; I'll fix.

File: .github/workflows/main_ci.yml (hunk 1 above)
Comment (burnpanck) [reply .github/workflows/main_ci.yml:5]: absolutely

File: include/foonathan/memory/allocator_storage.hpp [hunk 5]
```diff
@@ -126,7 +126,7 @@ namespace foonathan
                 FOONATHAN_REQUIRES(
                     (!std::is_base_of<allocator_storage, typename std::decay<Alloc>::type>::value))>
             allocator_storage(Alloc&& alloc,
-                              FOONATHAN_SFINAE(new storage_policy(detail::forward<Alloc>(alloc))))
+                              FOONATHAN_SFINAE(new storage_policy(std::declval<Alloc&&>())))
```
Comment (foonathan) [inline include/foonathan/memory/allocator_storage.hpp:129]: Nitpick: the && is not necessary, `std::declval` adds it already: https://en.cppreference.com/w/cpp/utility/declval

**Review by foonathan — CHANGES_REQUESTED** [review]

File: include/foonathan/memory/allocator_storage.hpp (hunk 5 above)
Comment (burnpanck) [reply include/foonathan/memory/allocator_storage.hpp]: Thx, I didn't know! I always thought that `std::declval` would return a value unless you gave it a reference type.

**Review by foonathan — APPROVED** [review]

**Comment by foonathan** [comment]
Thank you.

**Pull request merged**